node-red-node-wot/plugin-resources-src/node-wot-plugin-lib.ts

```diff
@@ -0,0 +1,954 @@
+        offsetY += Y_GAP
+    }
+    let flow = JSON.parse(flowStr)
+    return { flow, offsetY }
+}
+
+const makeCommonFlow = (commonGenIds, params, offsetY, existedUiBaseId) => {
+    let flowAndOffsetY = replaceParamsAndIds(THING_COMMON_TEMP, params, commonGenIds, null, null, null, offsetY)
+    // if ui-base node already existed on canvas, remove the node from new flow
+    let flow = flowAndOffsetY.flow.filter((node) => node.id !== existedUiBaseId)
+    flowAndOffsetY.flow = flow
+    return flowAndOffsetY
+}
+
+const makePropertyFlow = (commonGenIds, propertyGenIds, params, tdProperty, offsetY) => {
+    let flow = []
+    let flowAndOffsetY = replaceParamsAndIds(
+        PROPERTY_COMMON_TEMP,
+        params,
+        commonGenIds,
+        propertyGenIds,
+        null,
+        null,
+        offsetY
+    )
+    flow = flow.concat(flowAndOffsetY.flow)
+    if (!tdProperty.writeOnly) {
+        flowAndOffsetY = replaceParamsAndIds(
+            PROPERTY_READ_TEMP,
+            params,
+            commonGenIds,
+            propertyGenIds,
+            null,
+            null,
+            flowAndOffsetY.offsetY
+        )
+        flow = flow.concat(flowAndOffsetY.flow)
+        if (["integer", "number"].includes(tdProperty.type)) {
+            flowAndOffsetY = replaceParamsAndIds(
+                PROPERTY_READ_CHART_TEMP,
+                params,
+                commonGenIds,
+                propertyGenIds,
+                null,
+                null,
+                flowAndOffsetY.offsetY
+            )
+            flow = flow.concat(flowAndOffsetY.flow)
+        }
+    }
+    if (!tdProperty.readOnly) {
+        flowAndOffsetY = replaceParamsAndIds(
+            PROPERTY_WRITE_TEMP,
+            params,
+            commonGenIds,
+            propertyGenIds,
+            null,
+            null,
+            flowAndOffsetY.offsetY
+        )
+        flow = flow.concat(flowAndOffsetY.flow)
+    }
+    return { flow, offsetY: flowAndOffsetY.offsetY }
+}
+
+const makeActionFlow = (commonGenIds, actionGenIds, params, offsetY) => {
+    let flowAndOffsetY = replaceParamsAndIds(ACTION_TEMP, params, commonGenIds, null, actionGenIds, null, offsetY)
+    return flowAndOffsetY
+}
+
+const makeEventFlow = (commonGenIds, eventGenIds, params, offsetY) => {
+    let flowAndOffsetY = replaceParamsAndIds(EVENT_TEMP, params, commonGenIds, null, null, eventGenIds, offsetY)
+    return flowAndOffsetY
+}
+
+/**
+ * <copy from @node-red/util/lib/util.js>
+ *
+ * Generates a psuedo-unique-random id.
+ * @return {String} a random-ish id
+ * @memberof @node-red/util_util
+ */
+function generateId() {
+    var bytes = []
+    for (var i = 0; i < 8; i++) {
+        bytes.push(
+            //@ts-ignore
+            Math.round(0xff * Math.random())
+                .toString(16)
+                .padStart(2, "0")
+        )
+    }
+    return bytes.join("")
+}
+
+function checkBinding(tdStr: string, binding: string) {
```

Comment:
This is done in node-wot at https://github.com/eclipse-thingweb/node-wot/blob/master/packages/td-tools/src/td-helpers.ts#L27 a bit but in general, this kind of code should be in node-wot or td-tools

Reply:
Thank you for your guidance!

I will modify it to use td-tools for protocol detection.
I will also modify the following parts.
https://github.com/eclipse-thingweb/node-red/blob/main/node-red-node-wot/src/wot-thing.html#L6

Reply:
@egekorkan 

Since the checkBinding method you mentioned works on a browser, I would like to use the following `browser- bundle`. I am a little concerned about the size of the bundle, though, since it includes unnecessary modules.

https://cdn.jsdelivr.net/npm/@node-wot/browser-bundle/dist/wot-bundle.min.js

When I pass td as an argument and execute the Wot.Tools.findProtocol method, I get an error because td.base does not exist.
```
wot-bundle.min.js:70296 Uncaught TypeError: Cannot read properties of undefined (reading 'indexOf')
    at Object.findProtocol (wot-bundle.min.js:70296:28)
    at <anonymous>:1:11
```

The TD passed as an argument is an object of the exposed Thing Description, but is there a problem with its usage?

Reply:
@danielpeintner could you help a bit? I think that base should not be necessary but the code indeed assumes it. Ideally, I would prefer a code like @hidetak 's in node-wot

Reply:
> Since the checkBinding method you mentioned works on a browser, I would like to use the following browser- bundle. I am a little concerned about the size of the bundle, though, since it includes unnecessary modules.

This should be fixed later on after further restructuring. @relu91 @danielpeintner I think we can move these kind of helpers to the td-tools and create corresponding packages? It is required in node-wot too so it might be a more delicate balance of dependencies. At the same time, a protocol detection code should not change that often.

Reply:
I created https://github.com/eclipse-thingweb/node-wot/issues/1270 to track the issue and work on it.

Reply:
Ok given thumbs ups, let's merge :)

Reply:
Thank you for merging.
If you have any feedback on this feature, I would appreciate it.